### PR TITLE
network: fix segfault in manager_free()

### LINKED ISSUE #10605: systemd-networkd seems to be crashing sporadically
Reviewing https://github.com/systemd/systemd/pull/10542, I ran `systemd-networkd-tests.py` a few times. When run under `ASan`, `systemd-networkd` just crashed with something like
```
systemd-networkd[10250]: ASAN:SIGSEGV
systemd-networkd[10250]: ==10250== ERROR: AddressSanitizer: SEGV on unknown address 0x60620000f090 (pc 0x60620000f090 sp 0x7ffe17c9cfd8 bp 0x7ffe17c9d090 T0)
systemd-networkd[10250]: AddressSanitizer can not provide additional info.
```
`valgrind` is a little bit more helpful:
```
test1: Gained carrier
test1: Gained IPv6LL
Network interface NamePolicy= disabled on kernel command line, ignoring.
test1: Lost carrier
Stopping Network Service...
vex amd64->IR: unhandled instruction bytes: 0xFF 0xFF 0x0 0x0 0x0 0x0 0x0 0xC0 0x0 0x0
vex amd64->IR:   REX=0 REX.W=0 REX.R=0 REX.X=0 REX.B=0
vex amd64->IR:   VEX=0 VEX.L=0 VEX.nVVVV=0x0 ESC=NONE
vex amd64->IR:   PFX.66=0 PFX.F2=0 PFX.F3=0
==10874== Invalid read of size 1
==10874==    at 0x985AFB0: ???
==10874==    by 0x4F2A9FC: internal_hashmap_remove (hashmap.c:1386)
==10874==    by 0x12C8B4: set_remove (set.h:43)
==10874==    by 0x12CC1B: routing_policy_rule_free (networkd-routing-policy-rule.c:51)
==10874==    by 0x11F186: manager_free (networkd-manager.c:1455)
==10874==    by 0x11617F: manager_freep (networkd-manager.h:98)
==10874==    by 0x116CA2: main (networkd.c:14)
==10874==  Address 0x202516c889 is not stack'd, malloc'd or (recently) free'd
==10874==
==10874==
==10874== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==10874==  Access not within mapped region at address 0x202516C889
==10874==    at 0x985AFB0: ???
==10874==    by 0x4F2A9FC: internal_hashmap_remove (hashmap.c:1386)
==10874==    by 0x12C8B4: set_remove (set.h:43)
==10874==    by 0x12CC1B: routing_policy_rule_free (networkd-routing-policy-rule.c:51)
==10874==    by 0x11F186: manager_free (networkd-manager.c:1455)
==10874==    by 0x11617F: manager_freep (networkd-manager.h:98)
==10874==    by 0x116CA2: main (networkd.c:14)
==10874==  If you believe this happened as a result of a stack
==10874==  overflow in your program's main thread (unlikely but
==10874==  possible), you can try to increase the size of the
==10874==  main thread stack using the --main-stacksize= flag.
==10874==  The main thread stack size used in this run was 8388608.
==10874==
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #10605 and #10608.

cc @evverx. Please test this.

### CODE REVIEW

File: src/network/networkd-manager.c [hunk 1]
```diff
@@ -1451,8 +1451,10 @@ void manager_free(Manager *m) {
         while ((pool = m->address_pools))
                 address_pool_free(pool);
 
-        set_free_with_destructor(m->rules, routing_policy_rule_free);
-        set_free_with_destructor(m->rules_foreign, routing_policy_rule_free);
+        /* routing_policy_rule_free() access m->rules and m->rules_foreign.
+         * So, it is necessary to set NULL after the sets are freed. */
+        m->rules = set_free_with_destructor(m->rules, routing_policy_rule_free);
+        m->rules_foreign = set_free_with_destructor(m->rules_foreign, routing_policy_rule_free);
```
Comment (evverx) [inline src/network/networkd-manager.c:1457]: @yuwata thank you! I'm about to launch a VM to run the test there. In the meantime, I'm wondering why I couldn't reproduce the crash reliably. Judging by the code it should always crash. Looks like another mystery :-)

**Comment by yuwata** [comment]
@evverx I've added a fix for #10608. Please test this again. Thank you.

**Comment by evverx** [comment]
The leak seems to have gone but now I can see another one (for some unknown reason `ASan` is having trouble unwinding the stack; the most important bits are there anyway):
```sh
==1472==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 128 byte(s) in 1 object(s) allocated from:
    #0 0x7ff822d90c48 in malloc (/lib64/libasan.so.5+0xeec48)
    #1 0x55f49028eee3 in malloc_multiply ../src/basic/alloc-util.h:78
    #2 0x55f49028f386 in routing_policy_rule_new ../src/network/networkd-routing-policy-rule.c:20
    #3 0x55f49029269b in routing_policy_rule_add_internal ../src/network/networkd-routing-policy-rule.c:247
    #4 0x55f4902931cb in routing_policy_rule_add_foreign ../src/network/networkd-routing-policy-rule.c:307
    #5 0x55f49025dff0 in manager_rtnl_process_rule ../src/network/networkd-manager.c:837
    #6 0x7ff821fb8c3c  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb21c3c)
    #7 0x7ff821fb920e  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb2220e)
    #8 0x7ff821fb970f  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb2270f)
    #9 0x7ff821fbbe4e  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb24e4e)
    #10 0x7ff8220097b2  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb727b2)
    #11 0x7ff82200fbe7  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb78be7)
    #12 0x7ff822010b2c  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb79b2c)
    #13 0x7ff822010e55  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb79e55)
    #14 0x55f490250c21 in main ../src/network/networkd.c:132
    #15 0x7ff8201bb11a in __libc_start_main (/lib64/libc.so.6+0x2311a)
Indirect leak of 6 byte(s) in 1 object(s) allocated from:
    #0 0x7ff822cdd320 in strdup (/lib64/libasan.so.5+0x3b320)
    #1 0x55f490292650 in routing_policy_rule_add_internal ../src/network/networkd-routing-policy-rule.c:242
    #2 0x55f4902931cb in routing_policy_rule_add_foreign ../src/network/networkd-routing-policy-rule.c:307
    #3 0x55f49025dff0 in manager_rtnl_process_rule ../src/network/networkd-manager.c:837
    #4 0x7ff821fb8c3c  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb21c3c)
    #5 0x7ff821fb920e  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb2220e)
    #6 0x7ff821fb970f  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb2270f)
    #7 0x7ff821fbbe4e  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb24e4e)
    #8 0x7ff8220097b2  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb727b2)
    #9 0x7ff82200fbe7  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb78be7)
    #10 0x7ff822010b2c  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb79b2c)
    #11 0x7ff822010e55  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb79e55)
    #12 0x55f490250c21 in main ../src/network/networkd.c:132
    #13 0x7ff8201bb11a in __libc_start_main (/lib64/libc.so.6+0x2311a)
Indirect leak of 6 byte(s) in 1 object(s) allocated from:
    #0 0x7ff822cdd320 in strdup (/lib64/libasan.so.5+0x3b320)
    #1 0x55f4902925e5 in routing_policy_rule_add_internal ../src/network/networkd-routing-policy-rule.c:236
    #2 0x55f4902931cb in routing_policy_rule_add_foreign ../src/network/networkd-routing-policy-rule.c:307
    #3 0x55f49025dff0 in manager_rtnl_process_rule ../src/network/networkd-manager.c:837
    #4 0x7ff821fb8c3c  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb21c3c)
    #5 0x7ff821fb920e  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb2220e)
    #6 0x7ff821fb970f  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb2270f)
    #7 0x7ff821fbbe4e  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb24e4e)
    #8 0x7ff8220097b2  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb727b2)
    #9 0x7ff82200fbe7  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb78be7)
    #10 0x7ff822010b2c  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb79b2c)
    #11 0x7ff822010e55  (/home/vagrant/systemd/build/src/shared/libsystemd-shared-239.so+0xb79e55)
    #12 0x55f490250c21 in main ../src/network/networkd.c:132
    #13 0x7ff8201bb11a in __libc_start_main (/lib64/libc.so.6+0x2311a)
SUMMARY: AddressSanitizer: 140 byte(s) leaked in 3 allocation(s).
```

**Comment by evverx** [comment]
As far as I can tell, it happened the first time the test was run. All the subsequent runs were fine.

**Comment by evverx** [comment]
It seems that it would be safe to merge the second commit where the memory leak is fixed. The first one apparently fixes the crash but somehow introduces the leak pointed out above.

**Comment by yuwata** [comment]
@evverx Updated. But not tested yet.

**Comment by poettering** [comment]
lgtm

**Comment by evverx** [comment]
I'm pretty sure the test is going to pass eventually :-)

**Pull request merged**